src/state.rs

```diff
@@ -95,6 +96,10 @@ pub struct ResolvedDnsStore {
     // in a future with support for per-pod DNS resolv.conf settings we may need
     // to change this to a map from source workload uid to resolved IP addresses.
     by_hostname: HashMap<String, ResolvedDns>,
+
+    // sf_by_hostname is a map which store the hostname and its SingleFlight properties to provides
+    // a duplicate function call suppression mechanism for DNS resolving tasks.
+    sf_by_hostname: HashMap<String, SingleFlight>,
```

Comment:
Hi @nmittler, The reason why I put `sf_by_hostname` in `state` which is an `Arc<RwLock<ProxyState>>` field of `DemandProxyState` is that `sf_by_hostname` hashmap should be similar to `by_hostname` hashmap in `ResolvedDnsStore` to re-use the RwLock to read or write data in safe way in concurrency. So I think it could not be a member of `DemandProxyState`.

Reply:
I understand that there will have to be a lock, but it's probably fine if we have another lock for this. That's why I suggested a separate structure that just handles resolution.

Having this here is polluting the simplicity of ProxyState. SingleFlight and the whole resolution process is a feature of DemandProxyState.

Reply:
Hi @nmittler, thanks for more description here, I'm working on this and submit another commit later